src/AggregateFunctions/AggregateFunctionDistinctJSONPaths.cpp

```diff
@@ -0,0 +1,350 @@
+            ReadBufferFromMemory buf(value.data, value.size);
+            auto type = decodeDataType(buf);
+            /// We should not have Nulls here but let's check just in case.
+            if (!isNothing(type))
+                data[path].insert(type->getName());
+        }
+    }
+
+    void addWholeColumn(const ColumnObject & column, const std::unordered_map<String, String> & typed_paths_type_names)
+    {
+        for (const auto & [path, _] : column.getTypedPaths())
+            data[path].insert(typed_paths_type_names.at(path));
+        for (const auto & [path, dynamic_column] : column.getDynamicPathsPtrs())
+        {
+            /// Add dynamic path only if it has at least one non-null value.
+            /// getNumberOfDefaultRows for Dynamic column is O(1).
+            if (dynamic_column->getNumberOfDefaultRows() != dynamic_column->size())
+                dynamic_column->getAllTypeNames(data[path]);
+        }
+
+        /// Iterate over all paths in shared data and decode the data types.
+        const auto [shared_data_paths, shared_data_values] = column.getSharedDataPathsAndValues();
+        for (size_t i = 0; i != shared_data_paths->size(); ++i)
+        {
+            auto path = shared_data_paths->getDataAt(i).toString();
+            auto value = shared_data_values->getDataAt(i);
+            ReadBufferFromMemory buf(value.data, value.size);
+            auto type = decodeDataType(buf);
+            /// We should not have Nulls here but let's check just in case.
+            if (!isNothing(type))
+                data[path].insert(type->getName());
+        }
+    }
+
+    void merge(const AggregateFunctionDistinctJSONPathsAndTypesData & other)
+    {
+        for (const auto & [path, types] : other.data)
+            data[path].insert(types.begin(), types.end());
+    }
+
+    void serialize(WriteBuffer & buf) const
+    {
+        writeVarUInt(data.size(), buf);
+        for (const auto & [path, types] : data)
+        {
+            writeStringBinary(path, buf);
+            writeVarUInt(types.size(), buf);
+            for (const auto & type : types)
+                writeStringBinary(type, buf);
+        }
+    }
+
+    void deserialize(ReadBuffer & buf)
+    {
+        size_t paths_size, types_size;
+        readVarUInt(paths_size, buf);
+        if (paths_size > DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE)
+            throw Exception(ErrorCodes::TOO_LARGE_ARRAY_SIZE, "Too large array size (maximum: {}): {}", DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE, paths_size);
+
+        data.reserve(paths_size);
+        String path, type;
+        for (size_t i = 0; i != paths_size; ++i)
+        {
+            readStringBinary(path, buf);
+            readVarUInt(types_size, buf);
+            if (types_size > DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE)
+                throw Exception(ErrorCodes::TOO_LARGE_ARRAY_SIZE, "Too large array size (maximum: {}): {}", DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE, types_size);
```

Comment:
```suggestion
        if (paths_size > DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE)
            throw Exception(ErrorCodes::TOO_LARGE_ARRAY_SIZE, "Too large array size for paths (maximum: {}): {}", DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE, paths_size);

        data.reserve(paths_size);
        String path, type;
        for (size_t i = 0; i != paths_size; ++i)
        {
            readStringBinary(path, buf);
            readVarUInt(types_size, buf);
            if (types_size > DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE)
                throw Exception(ErrorCodes::TOO_LARGE_ARRAY_SIZE, "Too large array size for types (maximum: {}): {}", DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE, types_size);
```
Or something similar, so the two exceptions can be distinguished easily not just based on the row number in the source file.